pkg/scheduler/framework/preemption/preemption.go

```diff
@@ -203,17 +287,23 @@ func (ev *Evaluator) Preempt(ctx context.Context, pod *v1.Pod, m framework.NodeT
 		return nil, framework.NewStatus(framework.Unschedulable, "no candidate node for preemption")
 	}
 
+	logger.V(2).Info("the target node for the preemption is determined", "node", bestCandidate.Name(), "pod", klog.KObj(pod))
+
 	// 5) Perform preparation work before nominating the selected candidate.
-	if status := ev.prepareCandidate(ctx, bestCandidate, pod, ev.PluginName); !status.IsSuccess() {
-		return nil, status
+	if ev.enableAsyncPreemption {
+		ev.prepareCandidateAsync(bestCandidate, pod, ev.PluginName)
```

Comment:
In async version, preemption failures does not prevent scheduler from annotating pod as unschedulable and setting nominated node (in `handleSchedulingFailure`). Isn't there a risk that the pod will remain unschedulable forever?

Reply:
Good point, I implemented `PodActivator` to deal with the failure scenario; we move the preemptor to activeQ when the preemption goroutine fails

Reply:
Not sure yet how the activator works, but I was thinking about something simpler. What about setting error status in `preempting` which would be read the next time the PreEnque is called?

Maybe I'm missing something obvious, but I've just noticed that the plugin does not register to any cluster events. I don't see us calling `moveToActiveQ` either, so how was it supposed to be unblocked?

Reply:
> What about setting error status in preempting which would be read the next time the PreEnque is called?

The problem here is, for example, when the preemption API calls aren't successful at all, the preemptor Pod gets stuck in the unschedulable pod pool because no Pod/Delete event happens.
So, we have to somehow let the preemption plugin move the Pod out of the unsched pod pool. That's `Activate()`.

> Maybe I'm missing something obvious, but I've just noticed that the plugin does not register to any cluster events. I don't see us calling moveToActiveQ either, so how was it supposed to be unblocked?

We don't need to register any events in the preemption plugin.
For example, the pod is originally rejected by the node resource fit,
1. the preemptor pod gets `unschedulablePlugin: node-resource-fit`
2. the scheduling queue tries to requeue this pod every time node-resource-fit's QHint returns Queue, but the preemption plugin blocks the requeueing at PreEnqueue until all the preemption is done.
3. When the last pod is deleted, the scheduling queue, again, should tries to requeue the preemptor pod, and this requeueing goes through because the preemption plugin doesn't block this pod anymore at PreEnqueue.

Reply:
> We don't need to register any events in the preemption plugin because the preemptor Pod should have the other plugins in unschedulablePlugins.
> 
Isn't it something https://github.com/kubernetes/kubernetes/pull/127180 is going to change, so that only the GatedPlugin's events are supposed to be checked? Here we don't register any, but I think it's possible that preemption plugin will become the gated one. Note that `runPreEnqueuePlugins` can reassign to the preemption plugin, as it become a blocking one (for instance when resource-fit let the pod go for some other reason).

BTW. why we need `preempting` map then if we wait for 'resource-fit' plugin anyway?